service/history/ndc/history_replicator.go

```diff
@@ -555,6 +555,10 @@ func (r *HistoryReplicatorImpl) applyStartEvents(
 		)
 	}
 
+	if task.getVersionedTransition() != nil {
+		mutableState.InitTransitionHistory()
+	}
```

Comment:
This is the event based replication logic right? I think we should only do this for state-based tasks (sync state and backfill). Otherwise transition history logic will start to kick in even if the have the enableTransitionHistory flag disabled.

Reply:
This method is also used for state-based task (backfill).

```
BackfillHistoryEvents
	applyBackfillEvents
		applyBackfillEventsWithNew
			doApplyEvents(newTask)
				applyStartEvents
```

Reply:
sure, but my point is that transition history should not be set in the old stack. We probably need to do some refactoring or pass in extra context information here or init the transition history at a higher layer.

Reply:
updated using isStateBased flag.